### PR TITLE
<option> added children flatten, & warn if invalid

### LINKED ISSUE #2620: "Unable to find element" error on <option> elements
If the contents of the `<option>` element is constructed in special way, it breaks React on component update with "Unable to find element" error. Here's an example of `render` method body:

        var v1 = String(Math.random());
        var v2 = String(Math.random());
        return <select><option value={v2}>{v1} ({v2})</option></select>;

Here's working jsfiddle (with React 0.12.0): http://jsfiddle.net/rp3fz0nf/1/

Adding `key={v2}` to option props fixes the issue.

### PR DESCRIPTION
**Pull request by bloodyowl**
fixes #2620

### CODE REVIEW

File: src/browser/ui/dom/components/ReactDOMOption.js [hunk 1]
```diff
@@ -17,6 +17,7 @@ var ReactDOMSelect = require('ReactDOMSelect');
 var ReactElement = require('ReactElement');
 var ReactInstanceMap = require('ReactInstanceMap');
 var ReactPropTypes = require('ReactPropTypes');
+var ReactChildren = require('ReactChildren');
```
Comment (spicyj) [inline src/browser/ui/dom/components/ReactDOMOption.js]: Can you put this in alpha order?

File: src/browser/ui/dom/components/ReactDOMOption.js [hunk 2]
```diff
@@ -86,7 +87,22 @@ var ReactDOMOption = ReactClass.createClass({
       props = assign({}, props, {selected: this.state.selected});
     }
 
-    return option(props, this.props.children);
+    var content = '';
+
+    // Flatten children and warn if they aren't strings or numbers;
+    // invalid types are ignored.
+    ReactChildren.forEach(this.props.children, function(child) {
+      if (typeof child === 'string' || typeof child === 'number') {
```
Comment (spicyj) [inline src/browser/ui/dom/components/ReactDOMOption.js]: Can you make sure this skips over null/undefined children without a warning and add a test for that?

File: src/browser/ui/dom/components/__tests__/ReactDOMOption-test.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * @emails react-core
+ */
+
+'use strict';
+
+/*jshint evil:true */
+
+describe('ReactDOMOption', function() {
+  var React;
+  var ReactLink;
```
Comment (spicyj) [inline src/browser/ui/dom/components/__tests__/ReactDOMOption-test.js]: unused

**Comment by spicyj** [comment]
Just a few inlines – otherwise looks great. If you could make those updates and squash your commits I'll merge.

File: src/browser/ui/dom/components/ReactDOMOption.js (hunk 2 above)
Comment (syranide) [inline src/browser/ui/dom/components/ReactDOMOption.js]: boolean too I assume?

**Comment by bloodyowl** [comment]
@spicyj done :wink:

File: src/browser/ui/dom/components/ReactDOMOption.js [hunk 4]
```diff
@@ -86,7 +87,25 @@ var ReactDOMOption = ReactClass.createClass({
       props = assign({}, props, {selected: this.state.selected});
     }
 
-    return option(props, this.props.children);
+    var content = '';
+
+    // Flatten children and warn if they aren't strings or numbers;
+    // invalid types are ignored.
+    ReactChildren.forEach(this.props.children, function(child) {
+      if (child == null || child === false) {
```
Comment (syranide) [inline src/browser/ui/dom/components/ReactDOMOption.js]: `true` didn't get any love?

Comment (bloodyowl) [inline src/browser/ui/dom/components/ReactDOMOption.js]: whoops, fixed;

**Comment by MoOx** [comment]
really nice work !

File: src/browser/ui/dom/components/ReactDOMOption.js [hunk 5]
```diff
@@ -86,7 +87,25 @@ var ReactDOMOption = ReactClass.createClass({
       props = assign({}, props, {selected: this.state.selected});
     }
 
-    return option(props, this.props.children);
+    var content = '';
+
+    // Flatten children and warn if they aren't strings or numbers;
+    // invalid types are ignored.
+    ReactChildren.forEach(this.props.children, function(child) {
+      if (child == null || typeof child === 'boolean') {
```
Comment (spicyj) [inline src/browser/ui/dom/components/ReactDOMOption.js]: You shouldn't need the boolean check here; forEach should map bools to null for you. Can you try removing this? I think your test will still pass.

**Comment by bloodyowl** [comment]
@spicyj the test fails if I remove these lines.

```
>> not ok 256 - browser/ui/dom/components/__tests__/ReactDOMOption-test ReactDOMOption should ignore null/undefined/false children without warning.
# Expected 4 to be 0.
```

**Comment by bloodyowl** [comment]
done @spicyj

**Comment by spicyj** [comment]
Thanks!

**Pull request merged**